In [ ]:
import pandas as pd
import numpy as np

## Limpeza

Importando os reviews, organizando em ordem cronológica e removendo a informação de fuso horário. </br>
Remover o fuso é necessário para poder exportar os dados para o excel e usa-lo no Power BI.

In [ ]:
reviews = pd.read_excel('/content/Reviews_Bank.xlsx', parse_dates= ['Review Last Update Date and Time'])
reviews = reviews.sort_values('Review Last Update Date and Time')
reviews['Review Last Update Date and Time'] = reviews['Review Last Update Date and Time'].dt.tz_localize(None)
reviews.index = range(reviews.shape[0])
reviews

Package Name  ... Sentiment
0         B4 Bank  ...       NaN
1         B4 Bank  ...       NaN
2         B4 Bank  ...       NaN
3         B4 Bank  ...       NaN
4         B4 Bank  ...       NaN
...           ...  ...       ...
5674      B4 Bank  ...  Positive
5675      B4 Bank  ...  Positive
5676      B4 Bank  ...  Positive
5677      B4 Bank  ...  Positive
5678      B4 Bank  ...  Positive

[5679 rows x 9 columns]

## Tratamento

Verificando os dados que precisarão ser tratados

In [ ]:
pd.isnull(reviews).sum()

Package Name                           0
App Version Name                     256
Reviewer Language                      0
Device                                 6
Review Last Update Date and Time       0
Star Rating                            0
Review Title                        5678
Review Text                         2923
Sentiment                           4251
dtype: int64

  </br>

Entendendo o conteúdo das colunas para poder tratar os dados faltantes

In [ ]:
reviews['Sentiment'].unique()

array([nan, 'Positive', 'Negative', 'Neutral'], dtype=object)

In [ ]:
reviews['App Version Name'].unique()

array(['6.10.0', '6.10.1', nan, '6.7.6', '6.4.0', '6.9.8', '6.8.8',
       '6.9.9', '6.7.9', '6.8.7', '6.10.2', '5.3.1', '6.9.2', '6.9.3',
       '6.10.3', '6.7.8', '6.6.11', '6.5.7', '6.9.5', '6.8.4', '4.1.17',
       '6.9.1', '6.10.4', '6.10.5', '6.4.3', '6.8.9', '6.7.4', '6.9.7',
       '6.9.6', '4.1.7', '6.10.6', '6.8.5', '6.5.5', '6.6.8', '6.6.3',
       '6.7.1', '6.0.7', '6.5.10', '6.1.9', '5.0.7', '6.7.2', '6.1.4',
       '6.5.1', '6.9.0', '6.10.7', '4.1.8', '6.2.5', '4.1.14', '6.10.8',
       '6.10.9', '6.11.1'], dtype=object)

  </br>

Ao analisar a coluna Device deduzi que, pela pequena quantidade de dispositivos diferentes, poderia ter alguns que se destacassem na quantia de reviews e isso poderia denotar uma tendência

In [ ]:
reviews['Device'].unique()

array(['A4_Pro', 'starqltechn', 'htc_hiaeul', 'a9y18qlte', 'm253',
       'j7xelte', 'j7y17lte', 'dreamlte', 'ali', 'cepheus', 'lavender',
       'beyond2', 'dream2lte', 'ASUS_X008', 'jackpot2lte', 'sanders_nt',
       'a5y17lte', 'zenlte', 'deen_sprout', 'kane_sprout', 'j8y18lte',
       'on5xelte', 'platina', 'beryllium', 'star2qltechn', 'ASUS_X00QD',
       'crownqltechn', 'aljeter', 'potter_nt', 'ailsa_ii', 'beckham',
       'a7y17lte', 'j5y17lte', 'ocean_t', 'ASUS_X018_4', 'j5lte',
       'beyond1', 'j5xnlte', 'cedric', 'a7y18lte', 'montana',
       'jackpotlte', 'on7xelte', 'ML-TI0D-MS80X', 'j2y18lte', 'herolte',
       'ASUS_Z012D', 'OnePlus5T', 'beyond0', 'mlv5', 'dipper',
       'j4corelte', 'j4primelte', 'a70q', 'a5xelte', 'a50', 'albus',
       'HWCLT', 'daisy_sprout', 'channel', 'j4lte', 'cactus', 'jflte',
       'G3116', 'payton', 'a30', 'vince', 'klte', 'a9xproltesea', 'a10',
       'K33b36', 'greatlte', 'alicee', 'F5321', 'sailfish', 'montana_n',
       'rhannah', 'ASUS_

  </br>

Indo um pouco mais a fundo na coluna Devices, foi possível perceber que: </br>
*   O Device com a maior concentração representa 6,5% (230) de todos os Reviews (5679).
*   20% dos Devices(73) representam uma concentração de 84% dos Reviews.

</br>

Concluo que não existe concentração o bastante para fazer uma análise mais a fundo em um Device específico.

In [ ]:
reviews['Device'].value_counts()

starqltechn     230
dreamlte        215
on7xelte        210
star2qltechn    193
on5xelte        168
               ... 
ASUS_X00R_2       1
H3123             1
D5833             1
hero2qltetmo      1
panell_d          1
Name: Device, Length: 367, dtype: int64

In [ ]:
reviews['Device'].value_counts()[:74].sum()

4752

  </br>

Subistituindo os valores faltantes, seguindo o padrão por coluna:</br> </br>
Device -> Unknown </br>
Review Title/Text -> Blank</br>
Sentiment -> Emotionless</br>
App Version Name -> 0.0.0</br>
</br>
E criando uma coluna somente com a data do review.</br></br>

In [ ]:
reviews['Device'] = reviews['Device'].fillna('Unknown')
reviews['Review Title'] = reviews['Review Title'].fillna('Blank')
reviews['Review Text'] = reviews['Review Text'].fillna('Blank')
reviews['Sentiment'] = reviews['Sentiment'].fillna('Emotionless')
reviews['App Version Name'] = reviews['App Version Name'].fillna('0.0.0')
reviews['Review Last Update Date'] = reviews['Review Last Update Date and Time'].dt.date

In [ ]:
pd.isnull(reviews).sum()

Package Name                        0
App Version Name                    0
Reviewer Language                   0
Device                              0
Review Last Update Date and Time    0
Star Rating                         0
Review Title                        0
Review Text                         0
Sentiment                           0
Review Last Update Date             0
dtype: int64

In [ ]:
reviews

Package Name App Version Name  ...    Sentiment Review Last Update Date
0         B4 Bank           6.10.0  ...  Emotionless              2019-08-01
1         B4 Bank           6.10.0  ...  Emotionless              2019-08-01
2         B4 Bank           6.10.0  ...  Emotionless              2019-08-01
3         B4 Bank           6.10.0  ...  Emotionless              2019-08-01
4         B4 Bank           6.10.0  ...  Emotionless              2019-08-01
...           ...              ...  ...          ...                     ...
5674      B4 Bank           6.11.1  ...     Positive              2019-10-31
5675      B4 Bank           6.11.1  ...     Positive              2019-10-31
5676      B4 Bank           6.11.1  ...     Positive              2019-10-31
5677      B4 Bank           6.11.1  ...     Positive              2019-10-31
5678      B4 Bank           6.11.1  ...     Positive              2019-10-31

[5679 rows x 10 columns]

## Análise Primária
Algumas análises mais simples para conhecer o dataset. </br>

### Reviewer Language
Ao ver a variedade de línguas, decidi ir mais a fundo

In [ ]:
reviews['Reviewer Language'].unique()

array(['pt', 'en', 'fr', 'de', 'es', 'zh-Hans', 'it', 'nl', 'cs'],
      dtype=object)

In [ ]:
reviews['Reviewer Language'].value_counts()

pt         5461
en          188
es           15
de            6
fr            5
zh-Hans       1
cs            1
it            1
nl            1
Name: Reviewer Language, dtype: int64

 </br>

Olhando os Reviews de cada língua, com excessão do *en*, os textos são escritos em português mesmo que essa não seja a língua informada.</br>
No caso do *en*, olhando uma amostra de 10 Reviews(entre os 72 que escreveram algum texto) 20% foram escritos em inglês e o restante em *pt*.

In [ ]:
reviews[reviews['Reviewer Language'] == 'de']['Review Text']

125                                              Blank
1588                                             Blank
1737    Applicativo nunca falou em muitos anos de uso!
1917                                             Blank
2717                                             Blank
5206                       ok em geral mas muito lento
Name: Review Text, dtype: object

In [ ]:
reviews[reviews['Reviewer Language'] == 'fr']['Review Text']

82                                    app super eficiente
1706    Horrivel. Tentando acessar minha conta em viag...
3764                                                Blank
4322    O aplicativo deveria possibilitar ver os extra...
5204                                                Blank
Name: Review Text, dtype: object

In [ ]:
reviews.loc[(reviews['Reviewer Language'] == 'en') & (reviews['Review Text'] != 'Blank')]['Review Text']

32      Instável com token, esta cada vez pior! instav...
65                                            amazing app
96      Muito lento. Sujeito a flutuações inaceitáveis...
156     It has improved, it became less spotty and mor...
204     Bom aplicativo, mas não funciona a funçao de a...
                              ...                        
5519                                The bank account app!
5523    Aplicativo ainda é bem devagar... Precisa atua...
5537    Tudo ok. Programa de pontos acumulados do cart...
5623                                         Praticidade!
5658    I am able to perform remotely all bank operati...
Name: Review Text, Length: 72, dtype: object

### Star Rating
O App é muito bem ranqueado, somente a nota 5 representa 70% dos Reviews, combinando as notas 4 e 5 a soma alcança 86% do total.</br>

In [ ]:
reviews['Star Rating'].value_counts()

5    3966
4     917
1     372
3     261
2     163
Name: Star Rating, dtype: int64

In [ ]:
reviews['Star Rating'].value_counts()[:2].sum()

4883

 </br>


Analizando os comentários das notas 4 e 5 é possível destacar: </br>

*   44% (2144) dos Reviews tiveram algum comentário




In [ ]:
best_revs = reviews[reviews['Star Rating'] >= 4]
best_revs['Review Text'].value_counts()

Blank                         2739
Muito bom                       96
Excelente                       47
Ótimo                           45
Boa                             37
                              ... 
Muito prático e rápido.          1
prático                          1
Muito funcional. Aprovado!       1
Obrigada                         1
ótimo app !                      1
Name: Review Text, Length: 1535, dtype: int64

In [ ]:
best_revs['Review Text'].value_counts()[1:].sum()

2144

In [ ]:
best_revs['Review Text'].value_counts()[80:91]

Excepcional                2
muito bom obrigado         2
EXCELENTE.                 2
Eficiente.                 2
Facilita a vida            2
Bastante amigável          2
Rápido e prático           2
Gosto muito                2
Funciona perfeitamente.    2
Funciona muito bem         2
Ótimo atendimento          2
Name: Review Text, dtype: int64

 </br>

Analizando os comentários das notas 1 e 2 é possível destacar: </br>

*   Comentários em branco representam 18% (97) do total (535)

In [ ]:
low_revs = reviews[reviews['Star Rating'] <= 2]
low_revs['Review Text'].value_counts()

Blank                                                                                                                                                                                                                                                             97
Muito lento                                                                                                                                                                                                                                                        4
pessimo                                                                                                                                                                                                                                                            2
trava muito                                                                                                                                                                                                              

In [ ]:
low_revs['Review Text'].value_counts().sum()

535

In [ ]:
low_revs['Review Text'].value_counts()[1:].sum()

438

In [ ]:
low_revs['Review Text'].value_counts()[20:31]

Muito ruim e lento                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            1
Péssimo! Trava em consultas básicas e até no logon. Poucas funcionalidades quando comparado com as opções do site.                                                                                                                                                                                                                                                                                                                                                                                      

 </br>

### Reviews Text
Para poder analisar o texto de maneira mais objetiva, será feita uma contagem de palavras usadas. </br>
O processo começará agrupando-os por Star Rating, limpando a pontuação e deixando todas as palavras minusculas.

In [ ]:
def text_cleaner(text):
  text = text.replace(".", " ")
  text = text.replace(",", " ")
  text = text.replace("?", " ")
  text = text.replace("!", " ")
  text = text.replace("/", " ")
  text = text.replace("=", " ")
  text = text.replace(":", " ")
  text = text.replace("-", " ")
  text = text.replace("_", " ")
  return text

In [ ]:
text = reviews.groupby('Star Rating')['Review Text'].apply(' '.join)
text = text.map(text_cleaner)

 </br>

Após todos os comentários estarem agrupados, eles serão separados por nota e será contado quantas vezes cada palavra aparece.
Essa informação pode ajudar a filtrar os comentários mais importantes e fornecer ao cliente uma forma de pesquisa de mercado. 

Star Rating 1

In [ ]:
text1 = text[1]
text1 = text1.lower()

word_list1 = text1.split()
word_dict1 = {}

for word in word_list1: 
    word_dict1[word] = word_dict1.get(word, 0) + 1

word_count1 = pd.DataFrame.from_dict(word_dict1, orient= 'index')
word_count1.reset_index(level=0, inplace=True)
word_count1.columns = ['word', 'count']
word_count1.sort_values('count', ascending= False, inplace= True)
word_count1.reset_index(drop= True, inplace= True)
word_count1.head()

word  count
0    o    270
1    e    219
2   de    210
3    a    181
4  não    174

In [ ]:
word_count1[word_count1['count'] > 35]

word  count
0             o    270
1             e    219
2            de    210
3             a    181
4           não    174
5           que    146
6          para    100
7           app     97
8            do     94
9            no     69
10        blank     68
11   aplicativo     63
12        muito     63
13           um     60
14            é     58
15           b4     51
16          com     48
17  atualização     43
18      consigo     42
19          uma     40
20           em     39
21          nao     39
22           na     38
23         bank     37
24         mais     36

 </br>

Star Rating 2

In [ ]:
text2 = text[2]
text2 = text2.lower()

word_list2 = text2.split()
word_dict2 = {}

for word in word_list2: 
    word_dict2[word] = word_dict2.get(word, 0) + 1

word_count2 = pd.DataFrame.from_dict(word_dict2, orient= 'index')
word_count2.reset_index(level=0, inplace=True)
word_count2.columns = ['word', 'count']
word_count2.sort_values('count', ascending= False, inplace= True)
word_count2.reset_index(drop= True, inplace= True)
word_count2.head()

word  count
0    o    128
1   de    121
2    e     82
3    a     70
4  não     65

 </br>

Star Rating 3

In [ ]:
text3 = text[3]
text3 = text3.lower()

word_list3 = text3.split()
word_dict3 = {}

for word in word_list3: 
    word_dict3[word] = word_dict3.get(word, 0) + 1

word_count3 = pd.DataFrame.from_dict(word_dict3, orient= 'index')
word_count3.reset_index(level=0, inplace=True)
word_count3.columns = ['word', 'count']
word_count3.sort_values('count', ascending= False, inplace= True)
word_count3.reset_index(drop= True, inplace= True)
word_count3.head()

word  count
0     de    140
1      o    113
2      e     94
3  blank     87
4      a     79

 </br>

Star Rating 4

In [ ]:
text4 = text[4]
text4 = text4.lower()

word_list4 = text4.split()
word_dict4 = {}

for word in word_list4: 
    word_dict4[word] = word_dict4.get(word, 0) + 1

word_count4 = pd.DataFrame.from_dict(word_dict4, orient= 'index')
word_count4.reset_index(level=0, inplace=True)
word_count4.columns = ['word', 'count']
word_count4.sort_values('count', ascending= False, inplace= True)
word_count4.reset_index(drop= True, inplace= True)
word_count4.head()

word  count
0  blank    505
1     de    220
2      o    162
3      e    101
4      a     99

 </br>

Star Rating 5

In [ ]:
text5 = text[5]
text5 = text5.lower()

word_list5 = text5.split()
word_dict5 = {}

for word in word_list5: 
    word_dict5[word] = word_dict5.get(word, 0) + 1

word_count5 = pd.DataFrame.from_dict(word_dict5, orient= 'index')
word_count5.reset_index(level=0, inplace=True)
word_count5.columns = ['word', 'count']
word_count5.sort_values('count', ascending= False, inplace= True)
word_count5.reset_index(drop= True, inplace= True)
word_count5.head()

word  count
0  blank   2234
1  muito    570
2      e    448
3    bom    323
4     de    306

In [ ]:
word_count5[word_count5['count'] > 50]


word  count
0        blank   2234
1        muito    570
2            e    448
3          bom    323
4           de    306
5      prático    199
6            o    189
7    excelente    188
8          app    181
9        fácil    175
10       ótimo    153
11  aplicativo    129
12           a    127
13         que    120
14      rápido     94
15         boa     94
16          do     80
17        usar     79
18           é     77
19        para     70
20        tudo     62
21         não     60
22         bem     59
23          no     58
24      melhor     57
25      sempre     55
26       banco     54
27          as     53
28          b4     53
29   eficiente     52
30        bank     52
31         com     51
32        mais     51

##Exportação da Base
Exportando em CSV para caso seja usado para modelagem e em EXCEL para uso no Power BI

In [ ]:
reviews.to_csv('Cleaned_Reviews_Bank.csv', index= False)
reviews.to_excel('Cleaned_Reviews_Bank.xlsx', index= False)

## Proximos passos
- Análisar quais são os dispositivos com maior concentração de comentários negativos, caso seja um número expressivo, passar a informação para o cliente. 